ansible/script/keystone.sh

```diff
@@ -96,6 +96,58 @@ gelato_conf() {
     sed -i "s,OS_PASSWORD=.*$,OS_PASSWORD=$STACK_PASSWORD," $compose_file
 }
 
+keystone_credentials () {
+    export OS_AUTH_URL="http://${HOST_IP}/identity"
+    export OS_USERNAME=admin
+    export OS_PASSWORD="${STACK_PASSWORD}"
+    export OS_PROJECT_NAME=admin
+    export OS_USER_DOMAIN_NAME=Default
+    export OS_IDENTITY_API_VERSION=3
+}
+
+wait_for_keystone () {
+    local count=0
+    local interval=${1:-10}
+    local times=${2:-12}
+
+    while true
+    do
+        # get a token to check if keystone is working correctly or not.
+        # keystone credentials such as OS_USERNAME must be set before.
+        openstack token issue &> /dev/null
+        if [ "$?" == "0" ]; then
+            return
+        fi
+        count=`expr ${count} \+ 1`
+        if [ ${count} -ge ${times} ]; then
+            echo "ERROR: keystone didn't come up. Aborting..."
+            exit 1
+        fi
+        sleep ${interval}
+    done
+}
+
+update_keystone_endpoints () {
+    # update keystone endpoint URL for all the discovered interfaces
+    # using the same given URL.
+    # FIXME(thatsdone): better to make it as single endpoint updation?
+    local REGION=$1
+    local SERVICE=$2
+    local URL=$3
+    local SERVICE_ID
+    local INTERFACE
+
+    while read -r line
+    do
+        IFS=" " read -r -a array <<< ${line}
+        SERVICE_ID=${array[0]}
+        INTERFACE=${array[1]}
+        openstack endpoint set --region ${REGION} --service ${SERVICE} --interface ${INTERFACE} --url "${URL}" ${SERVICE_ID} 2>/dev/null
+
+    done < <(openstack endpoint list --service ${SERVICE} -f value 2>/dev/null | awk '{print $1, $6;}')
+
```

Comment:
Im not sure if there is a space between < and <, is it right here?

Reply:
Ah, this is a bash process substitution syntax. There is a space between the two '<'s. 
The first '<' means redirect, and the second '<' is actually a part of '<(command sequence)'. 
We need one space between them so that bash can distinguish '<<' or '<' + '<(command sequence)'.
I want to read 'openstack endpoint list' command output line by line there.

As for process substitution, please see: https://www.tldp.org/LDP/abs/html/process-sub.html

Reply:
OK, it's clear